base/applications/rapps/include/rosui.h

```diff
@@ -21,31 +25,34 @@ class CPointerArray
         m_hDpa = DPA_Create(GrowthRate);
     }
 
-    ~CPointerArray()
+    virtual ~CPointerArray()
     {
+        // Here, s_OnRemoveItem() can only call our own OnRemoveItem() implementation...
+        // unless we use the CRTP construct...
```

Comment:
OK here and below (see `s_OnRemoveItem()` and the `class CUiCollection` implementation), we face the problem that, at the time the `~CPointerArray()` destructor is called, the `s_OnRemoveItem()` helper (and then, the `OnRemoveItem() method) being called will **never ever** call any overridden implementation of `OnRemoveItem()`, because the derived class(es) have already been destructed.

(A).
I think here, the idea was to have any derived classes from the template-specialized `CPointerArray` to have a custom "object" deleter. (I would like to know @gigaherz opinion on this.)
I'm not sure this is of great use, unless there is also some sort of callback during the `Set/Insert/Append` calls to balance this. My point being, at least as far as this class (and the `CUiCollection` that derives from it) is concerned in RAPPS, UI objects are first allocated and stored separately, then pointer to them are added into this pointer-array, but, these objects are later manually deleted one after the other in RAPPS code implementation. And then the `CUiCollection` (and therefore the underlying `CPointerArray`) instance(s) get destructed, and at this point we don't really need the "collection"/array to do the deletion for us (see more on that later in this comment).

(B).
In order to have the `CPointerArray` template class be able, from this `s_OnRemoveItem` helper, to call the overridden implementation from the derived class (`CUiCollection` in our case), the only simple technique I've found is this so-called _Curiously recurring template pattern_ (CRTP), aka. what the ATL classes do, to specify as the first template "parameter" the class being derived, so that the specialization of the base class "knows" what derived class (and its methods) to access.
This is the reason of the (currently commented-out) `TBase` parameter you can see above in the definition, and the commented `TDeriv* pThis = static_cast<TDeriv*>(context);` line in `s_OnRemoveItem()`.
Then, `CUiCollection` needs to be defined instead as:
```c++
class CUiCollection :
    public CPointerArray<CUiCollection, CUiPrimitive>
```
and the whole construct will work: the overridden `CUiCollection::OnRemoveItem()` helper will be called during destruction.

However, the currently suggested implementation of `CUiCollection::OnRemoveItem()` (to explicitly `delete` the passed pointer) will be inadequate, due to the reason I've given in point (A). Indeed, with the modifications explained above, if we keep the current implementation of deleting the passed pointers, then we trigger a crash, since we would delete a pointer to a window object, that has already been separately deleted by some RAPPS cleanup code...

Reply:
If you wanna refactor, nuke the `CPointerArray` and use either `CSimpleArray` or `CArray`,
both should delete their items.
The whole DPA_XXX stuff is a thing you would use when writing in C, and you don't want to write too much boilerplate.